peer/gossip.go

```diff
@@ -0,0 +1,140 @@
+
+	gossip := func(ctx context.Context, subnet id.Hash, contentID []byte) error {
+		msg := wire.Msg{Version: wire.MsgVersion1, Type: wire.MsgTypePush, Data: contentID}
+
+		var chainedError error = nil
+		var receivers []id.Signatory
+		if bytes.Compare(subnet[:], GlobalSubnet[:]) == 0 {
+			receivers = addressTable.All()
+		} else {
+			receivers = addressTable.Subnet(subnet)
+		}
+		
+		for _, sig := range receivers {
+			addr, ok := addressTable.PeerAddress(sig)
+			if !ok {
+				logger.Error("gossip", zap.String("table", "peer not found"))
+				continue
+			}
+			if err := t.Send(ctx, sig, addr, msg); err != nil {
+				if chainedError == nil {
+					chainedError = fmt.Errorf("%v, gossiping to peer %v : %v", chainedError, sig, err)
+				} else {
+					chainedError = fmt.Errorf("gossiping to peer %v : %v", sig, err)
+				}
+			}
+		}
+		return chainedError
+	}
+	didReceivePush := func(from id.Signatory, msg wire.Msg) {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		addr, ok := addressTable.PeerAddress(from)
+		if !ok {
+			logger.Error("gossip", zap.String("table", "peer not found"))
+			return
+		}
+
+		hash := [32]byte{}
+		copy(hash[:], msg.Data[:])
+
+		if _, ok := contentResolver.Content(hash); !ok {
+			response := wire.Msg{
+				Version: wire.MsgVersion1,
+				Type: wire.MsgTypePull,
+				Data: hash[:],
+			}
+
+			contentResolver.Insert(hash, nil)
+			if err := t.Send(ctx, from, addr, response); err != nil {
+				contentResolver.Delete(hash)
+				logger.Error("gossip", zap.NamedError("pull", err))
+				return
+			}
+
+		}
+	}
+	didReceivePull := func(from id.Signatory, msg wire.Msg) {
+		hash := [32]byte{}
+		copy(hash[:], msg.Data[:])
+		if data, ok := contentResolver.Content(hash); ok {
+			response := wire.Msg{
+				Version: wire.MsgVersion1,
+				Type: wire.MsgTypeSync,
+				Data: data,
+			}
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			addr, ok := addressTable.PeerAddress(from)
+			if !ok {
+				logger.Error("gossip", zap.String("table", "peer not found"))
+				return
+			}
+			if err := t.Send(ctx, from, addr, response); err != nil {
+				logger.Error("gossip", zap.NamedError("pull", err))
+			}
+			return
+		}
+
+		logger.Error("gossip", zap.String("pull request", "data not present"))
+	}
+	didReceiveSync := func(from id.Signatory, msg wire.Msg) {
+		hash := id.NewHash(msg.Data)
+		content, ok := contentResolver.Content(dht.ContentID(hash))
+		if !ok {
+			logger.Error("gossip", zap.String("sync", "unknown id in sync message"))
+			return
+		}
+		if content != nil {
+			// TODO : Add debugging message for logger
+			return
+		}
+		contentResolver.Insert(dht.ContentID(hash), msg.Data)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		if err := gossip(ctx, id.Hash{}, hash[:]); err != nil {
```

Comment:
The empty hash should be used as a subnet identifier. This should be set to `msg.To`.